File: Learn_Basic/src/learn_py/learn_py/service_opencv/opencv_find_point_server.py
import cv2
import numpy as np
import rclpy
from rclpy.node import Node
from learn_interface.srv import TopicOpencv
from geometry_msgs.msg import Point
from sensor_msgs.msg import Image
from cv_bridge import CvBridge

def Find_Point(origin):
    # 将图像从BGR转换为HSV颜色空间
    hsv = cv2.cvtColor(origin, cv2.COLOR_BGR2HSV)

    # 定义红色的HSV范围
    hsv_lower = np.array([156, 43, 46])
    hsv_upper = np.array([180, 255, 255])
    mask = cv2.inRange(hsv, hsv_lower, hsv_upper)
    kernel = np.ones((2, 2), np.uint8)
    mask_opened = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
    contours, _ = cv2.findContours(mask_opened, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)

    # 处理查找结果
    radius_max = 0
    point_result = []
    for cnt in contours:
        epsilon = 0.1 * cv2.arcLength(cnt, True)
        approx = cv2.approxPolyDP(cnt, epsilon, True)
        # 拟合
        (x, y), radius = cv2.minEnclosingCircle(approx)
        if radius_max < radius:
            radius_max = radius
            point_result.append((int(x), int(y), int(radius)))  # 添加半径到结果中
            
    return radius_max != 0, point_result


def draw_point_circle(image, posArray):
    for pos in posArray:
        center = (pos[0], pos[1])
        radius = pos[2]  # 使用传入的半径
        cv2.circle(image, center, radius, (255, 0, 0), 2)
        
class PublisherNode(Node):
    def __init__(self, name, cap):
        super().__init__(name)
        self.server = self.create_service(TopicOpencv, "learn_service_opencv_find_point", self.handle_service_msg)
        self.bridge = CvBridge()
        self.cap = cap
        
    def handle_service_msg(self, request, response):
        self.get_logger().info(f'收到请求: scan_once={request.scan_once}')
        
        data = [Image(),[]]
        
        # 根据请求处理
        if request.scan_once:
            data = self.scan_camera()
            
        return self.echo(data, response=response)
    
    def scan_camera(self):
        ret, frame = self.cap.read()
        
        sendmessage = [frame,[]]
        
        if not ret:
            ret
        try:
            isFind, points = Find_Point(frame)
            
            pub_points = []
            
            for point in points:
                if(point[2] > 10):
                    pub_points.append(point)
            
            sendmessage[1] = pub_points
            
            if isFind:
                draw_point_circle(frame, pub_points)

        except Exception as e:
            print(e)  # 打印异常信息
        
        finally:
            return sendmessage
        
    def echo(self, data, response):
        response.image = self.bridge.cv2_to_imgmsg(data[0], encoding="bgr8")
        
        # 转换点数据
        point_msgs = []
        for point in data[1]:
            p_msg = Point()
            p_msg.x = float(point[0])
            p_msg.y = float(point[1])
            p_msg.z = float(point[2])
            point_msgs.append(p_msg)
        
        response.points = point_msgs
        return response

def main(args = None):
    
    rclpy.init(args=args)
    
    cap = cv2.VideoCapture(0)
    
    server = PublisherNode("opencv_find_point_service_node", cap)
    
    try:
        rclpy.spin(server)
    except KeyboardInterrupt:
        pass
    finally:
        cap.release()
        server.destroy_node()
        rclpy.shutdown()

File: Learn_Basic/src/learn_cmake/src/service_opencv/opencv_find_point_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "learn_interface/srv/topic_opencv.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.hpp"

#include <opencv2/opencv.hpp>
#include <vector>
#include <tuple>
#include <memory>

// 查找红色点
std::pair<bool, std::vector<std::tuple<int,int,int>>> Find_Point(const cv::Mat &origin)
{
    cv::Mat hsv;
    cv::cvtColor(origin, hsv, cv::COLOR_BGR2HSV);

    cv::Scalar hsv_lower(156, 43, 46);
    cv::Scalar hsv_upper(180, 255, 255);
    cv::Mat mask;
    cv::inRange(hsv, hsv_lower, hsv_upper, mask);

    cv::Mat kernel = cv::Mat::ones(2, 2, CV_8U);
    cv::Mat mask_opened;
    cv::morphologyEx(mask, mask_opened, cv::MORPH_OPEN, kernel);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask_opened, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_NONE);

    int radius_max = 0;
    std::vector<std::tuple<int,int,int>> point_result;

    for (auto &cnt : contours)
    {
        double epsilon = 0.1 * cv::arcLength(cnt, true);
        std::vector<cv::Point> approx;
        cv::approxPolyDP(cnt, approx, epsilon, true);

        cv::Point2f center;
        float radius;
        cv::minEnclosingCircle(approx, center, radius);

        if (radius_max < radius)
        {
            radius_max = static_cast<int>(radius);
            point_result.push_back({static_cast<int>(center.x), static_cast<int>(center.y), static_cast<int>(radius)});
        }
    }

    return {radius_max != 0, point_result};
}

// 在图像上画圈
void draw_point_circle(cv::Mat &image, const std::vector<std::tuple<int,int,int>> &posArray)
{
    for (auto &pos : posArray)
    {
        cv::Point center(std::get<0>(pos), std::get<1>(pos));
        int radius = std::get<2>(pos);
        cv::circle(image, center, radius, cv::Scalar(255, 0, 0), 2);
    }
}

class ServiceNode : public rclcpp::Node
{
public:
    ServiceNode(const std::string & name)
    : Node(name)
    {
        server_ = this->create_service<learn_interface::srv::TopicOpencv>(
            "learn_service_opencv_find_point",
            std::bind(&ServiceNode::handle_service_msg, this,
                      std::placeholders::_1, std::placeholders::_2));
        cap_.open(0);
        if (!cap_.isOpened())
        {
            RCLCPP_ERROR(this->get_logger(), "Error opening video capture");
        }
    }

private:
    void handle_service_msg(
        const std::shared_ptr<learn_interface::srv::TopicOpencv::Request> request,
        std::shared_ptr<learn_interface::srv::TopicOpencv::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "收到请求: scan_once=%d", request->scan_once);

        cv::Mat frame;
        cap_ >> frame;
        if (frame.empty())
        {
            RCLCPP_ERROR(this->get_logger(), "摄像头读取失败");
            return;
        }

        try
        {
            auto [isFind, points] = Find_Point(frame);

            std::vector<std::tuple<int,int,int>> pub_points;
            for (auto &point : points)
            {
                if (std::get<2>(point) > 10)
                    pub_points.push_back(point);
            }

            if (isFind)
            {
                draw_point_circle(frame, pub_points);
            }

            // 转换图像
            response->image = *cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", frame).toImageMsg();

            // 转换点数据
            std::vector<geometry_msgs::msg::Point> point_msgs;
            for (auto &point : pub_points)
            {
                geometry_msgs::msg::Point p;
                p.x = static_cast<double>(std::get<0>(point));
                p.y = static_cast<double>(std::get<1>(point));
                p.z = static_cast<double>(std::get<2>(point));
                point_msgs.push_back(p);
            }
            response->points = point_msgs;
        }
        catch (const cv::Exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "OpenCV异常: %s", e.what());
        }
    }

    rclcpp::Service<learn_interface::srv::TopicOpencv>::SharedPtr server_;
    cv::VideoCapture cap_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ServiceNode>("opencv_find_point_service_node");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
